### DISCUSSION TITLE
Bzlmod: Poll for the most wanted modules in the Bazel Central Registry

### DISCUSSION THREAD

**Discussion by meteorcloudy** [bazelbuild/bazel#18329] [General]
Hello Bazel community! 

Based on recent feedback on Bzlmod from different channels, while many of you are enjoying the new package manager for Bazel, there are still many users facing challenges while trying to migrate to Bzlmod. One of the main issues seems to be the limited number of dependencies available in the [Bazel Central Registry](https://github.com/bazelbuild/bazel-central-registry).

To accelerate the Bzlmod migration process, we would like to know which dependencies you want to see in the Bazel Central Registry. Your input will help us and the community understand which rule sets and libraries are the most important ones to migrate.

Since GitHub Discussion poll doesn't support [multiple choices](https://github.com/orgs/community/discussions/52039) yet, we are doing the poll in the following way:

- Please upvote the corresponding replies below for your most wanted dependencies. 
- If your wanted dependency isn't listed or still missing functionalities, please create a new reply with the name of the dependency with some context.
- Please link to a [module wanted issue](https://github.com/bazelbuild/bazel-central-registry/issues?q=is%3Aissue+is%3Aopen+label%3A%22module+wanted%22) if there is one (and feel free to create one).

If you have any further concerns or suggestions regarding Bzlmod migration, feel free to share them in the comments as well!

Thank you for your participation and support!

### Useful links

If you are new to Bzlmod, here are some useful links:

- Official doc: https://bazel.build/external/overview#bzlmod 
- [Bzlmod Migration Guide](https://bazel.build/external/migration)
- Minimal examples: https://github.com/bazelbuild/examples/tree/main/bzlmod

**Comment by meteorcloudy** [comment]
[protobuf](https://github.com/bazelbuild/bazel-central-registry/issues/320)

As of May 2023, the BCR does have the [protobuf module ](https://github.com/bazelbuild/bazel-central-registry/tree/main/modules/protobuf) providing minimal Bzlmod support with some patches applied. But there are still a lot to improve and it should be better maintained.

**Reply by chrislovecnm** [reply]
rules_kotlin is blocking this migration.

**Reply by shahms** [reply]
The version present in the BCR is too old for us (Kythe).

**Comment by fmeum** [comment]
[rules_kotlin](https://github.com/bazelbuild/bazel-central-registry/issues/529)

**Reply by chrislovecnm** [reply]
https://github.com/bazelbuild/rules_kotlin/pull/900 and https://github.com/bazelbuild/rules_kotlin/pull/899 are the start of the work.  I am not certain if these PRs will provide feature parity.  Those two PRs are also quite old.

**Reply by chrislovecnm** [reply]
Do we have a status on this?

**Comment by jsharpe** [comment]
https://github.com/pybind/pybind11_bazel

**Reply by lalten** [reply]
This one is in BCR but https://github.com/pybind/pybind11_bazel/issues/66 means it will only work with non-hermetic local Python.

**Reply by junyer** [reply]
This is resolved now. :)

**Comment by jsharpe** [comment]
https://github.com/grailbio/bazel-toolchain

**Reply by jsharpe** [reply]
https://github.com/jsharpe/bazel-toolchain/tree/bzlmod - I started on this some time ago but can't remember how far I got with getting this working...

**Reply by jsharpe** [reply]
https://github.com/grailbio/bazel-toolchain/pull/198 - this PR will eventually fix this; some help is required in getting the tests all running when bzlmod is enabled.

**Comment by jsharpe** [comment]
Annotation support for pip_parse extension in rules_python

**Reply by chrislovecnm** [reply]
Fixed.  Next release we will have full feature parity

**Reply by chrislovecnm** [reply]
@rickeylev I think we can say that we are fully compatible with bzlmod now.  Please correct me otherwise.

**Reply by rickeylev** [reply]
Yeah, I think we do.

**Comment by brentleyjones** [comment]
Stardoc

**Reply by meteorcloudy** [reply]
Thanks to @fmeum, Stardoc is close https://github.com/bazelbuild/stardoc/pull/141

**Reply by brentleyjones** [reply]
And now with 0.5.4 it works!

**Reply by chrislovecnm** [reply]
Are we still missing the capability for stardoc to generate bzlmod extension docs?

**Reply by Wyverald** [reply]
> Are we still missing the capability for stardoc to generate bzlmod extension docs?

Yes; @tetromino has a change in flight to add that.

**Reply by tetromino** [reply]
> Yes; @tetromino has a change in flight to add that.

A series of changes: one in flight + approx. 2 followups in my head.

**Comment by sluongng** [comment]
[Rules_rust](https://github.com/bazelbuild/bazel-central-registry/issues/329)

**Reply by chrislovecnm** [reply]
https://github.com/bazelbuild/rules_rust/issues/1493 is tracking this issue

**Reply by csmulhern** [reply]
rules_rust currently has no maintainer with Bzlmod experience. If anyone could help with contributions / reviews, that would be much appreciated!

**Reply by fmeum** [reply]
> rules_rust currently has no maintainer with Bzlmod experience. If anyone could help with contributions / reviews, that would be much appreciated!

Happy to help with reviews, just mention me.

**Reply by csmulhern** [reply]
Will do, thanks Fabian!

**Reply by Wyverald** [reply]
I have a pair-PR-review session with @scentini tomorrow morning :)

**Comment by jwbee** [comment]
[Google benchmark](https://github.com/bazelbuild/bazel-central-registry/issues/384)

**Reply by jsharpe** [reply]
This is in at HEAD for the c++ bindings: https://github.com/google/benchmark/blob/main/MODULE.bazel with the caveat that its using an older rules_python extension API so will break in the future; this can be fairly easily fixed though when the next version of rules_python comes out.

**Reply by meteorcloudy** [reply]
Thanks to @macandy13, google benchmark is added in https://github.com/bazelbuild/bazel-central-registry/pull/795

**Comment by adam-singer** [comment]
[bazelbuild/rules_scala](https://github.com/bazelbuild/bazel-central-registry/issues/522)

**Reply by chrislovecnm** [reply]
https://github.com/bazelbuild/rules_scala/issues/1482 is tracking this issue.  No work has started as of yet.

**Reply by mbland** [reply]
https://github.com/bazelbuild/rules_scala/issues/1482 is in the home stretch now.

**Reply by mbland** [reply]
Should've posted here months ago, but the [rules_scala BCR module](https://registry.bazel.build/modules/rules_scala) has existed since 2025-05-09.

**Comment by adam-singer** [comment]
[rules_license](https://github.com/bazelbuild/rules_license)

**Reply by chrislovecnm** [reply]
[It is in the repository.](https://registry.bazel.build/modules/rules_license)

**Comment by adam-singer** [comment]
[rules_pkg](https://github.com/bazelbuild/rules_pkg)

**Reply by brentleyjones** [reply]
I believe this is already on BCR?

**Reply by meteorcloudy** [reply]
Yes, @aiuto are there any functionalities that are not available with Bzlmod enabled?

**Reply by aiuto** [reply]
I have not broken out a separate module for the RPM toolchain yet.

**Comment by adam-singer** [comment]
[rules_antlr](https://github.com/marcohu/rules_antlr)

**Comment by adam-singer** [comment]
[rules_proto_grpc](https://github.com/rules-proto-grpc/rules_proto_grpc)

**Comment by adam-singer** [comment]
[rules_go](https://github.com/bazelbuild/rules_go)

**Reply by fmeum** [reply]
This is already on the BCR, with only a few features missing compared to WORKSPACE. See https://github.com/bazelbuild/rules_go/blob/master/docs/go/core/bzlmod.md.

**Reply by chrislovecnm** [reply]
@fmeum do you have feature parity now?

**Reply by fmeum** [reply]
Some niche features are missing, but some core features are actually more usable with Bzlmod than without. All in all, I would say this is done.

**Reply by chrislovecnm** [reply]
@fmeum do we have feature parity?

**Reply by fmeum** [reply]
Except for nogo I would say yes. Suggestions for how to model nogo with Bzlmod are very welcome.

**Comment by neal-codaio** [comment]
[rules_jsonnet](https://github.com/bazelbuild/rules_jsonnet)

**Comment by mortenmj** [comment]
[jsonnet](https://github.com/google/jsonnet) and [go-jsonnet](https://github.com/google/go-jsonnet)

**Reply by mortenmj** [reply]
This is done

**Reply by chris-codaio** [reply]
How does one use jsonnet via bzlmod?  I can't find any documentation.

**Comment by chrislovecnm** [comment]
[rules_dotnet](https://github.com/bazelbuild/rules_dotnet/issues/332) does not yet have support.

**Comment by coryan** [comment]
libcurl: https://github.com/curl/curl/

**Reply by sluongng** [reply]
This might be an easy one as both tensorflow and Kythe are maintaining a BUILD file for this https://github.com/kythe/kythe/blob/master/third_party/curl.BUILD

**Reply by coryan** [reply]
Seems like the Kythe and tensorflow set `CURL_CA_BUNDLE` incorrectly on some distros.

Maybe this version is a better place to start from: https://github.com/googleapis/google-cloud-cpp/blob/main/bazel/curl.BUILD

**Reply by aiuto** [reply]
I've talked with the curl maintainers about Bazel support. Daniel is OK with that, as long as someone else does the work. Since I maintain the curl BUILD file that google3 users, I'm probably the domain expert on that. Perhaps I can finally try to do this.
I don't know about the other BUILDs, but my hunch is that they are probably linux centric, which is not suffucient portabiliity.

**Reply by sluongng** [reply]
I don't think all modules need to have multi-platforms support from day 1.
If you could help opensource / provide Bazel setup for curl, I'm sure it will be updated overtime as folks adopt it to their needs.

**Reply by lalten** [reply]
@aiuto any progress on this one?

**Reply by Vertexwahn** [reply]
@lalten I did an experiment [here](https://github.com/Vertexwahn/third_party/tree/main/curl-7.69.1) to Bzmod curl 7.69.1. I simply copied https://github.com/googleapis/google-cloud-cpp/blob/main/bazel/curl.BUILD. 

bazel `build build //:curl` results in:

> Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
> In file included from lib/urldata.h:127,
>                  from lib/strerror.h:25,
>                  from lib/strerror.c:46:
> lib/vssh/ssh.h:25:10: fatal error: curl_setup.h: No such file or directory
>    25 | #include "curl_setup.h"
>       |          ^~~~~~~~~~~~~~

**Reply by Vertexwahn** [reply]
@lalten Could fix the issue

**Reply by Vertexwahn** [reply]
https://github.com/bazelbuild/bazel-central-registry/pull/1429

**Comment by Vertexwahn** [comment]
Boost

**Reply by sluongng** [reply]
You might be interested in https://github.com/bazelboost/registry

**Reply by Vertexwahn** [reply]
Does it already work?

**Reply by jsharpe** [reply]
Doesn't look like it - none of the modules have versions present

**Reply by zaucy** [reply]
I've been updating https://github.com/bazelboost/registry incrementally. It's come a long way since this post. If you're interested in using it please do and report issues you have. Not every module is ready, but for my projects it's working well.

**Comment by Vertexwahn** [comment]
Qt

**Reply by sluongng** [reply]
is there a rule for Qt already to add to the registry? Or is this a rules request?

**Reply by Vertexwahn** [reply]
This is a request - there are some WORKSPACE approaches in place:
- https://github.com/Vertexwahn/rules_qt6
- [aabtop/rules_qt](https://github.com/aabtop/rules_qt)
- [bbreslauer/qt-bazel-example](https://github.com/bbreslauer/qt-bazel-example)
- [justbuchanan/bazel_rules_qt](https://github.com/justbuchanan/bazel_rules_qt)
- https://github.com/kklochkov/rules_qt
There is also som approch to build Qt from Scratch using Qt that works with Linux and Bazel already - but can currently not find the repo...
They all lacking Bzlmod support...

**Comment by Vertexwahn** [comment]
https://github.com/bazelbuild/rules_webtesting

**Comment by Vertexwahn** [comment]
https://github.com/emscripten-core/emsdk/tree/main/bazel

**Comment by Vertexwahn** [comment]
TensorFlow

**Reply by lucien-perouze** [reply]
Yes please !

**Comment by sluongng** [comment]
https://github.com/google/copybara would be really nice.

The repo is tightly coupled with Bazel.git so I am very curious about how such a migration could be done.
It would be nice if this could onboard Bazel CI as well to discover breakages early since I had to apply several monkey patches locally to get it to build.

**Reply by sluongng** [reply]
This might be a cool move to double the Java implementation of starlark into a separate module. Which is the shared part between bazel.git and copybara.git.

**Comment by fausturs** [comment]
openssl 
cyrus-sasl
curl
arrow

**Reply by illicitonion** [reply]
FYI openssl is now hosted: https://registry.bazel.build/modules/openssl

**Comment by xFile3160** [comment]
halide
tensorflow-lite
opencv
ffmpeg
node-addon-api
node-api-headers
opencl-clhpp
opencl-headers
backward-cpp

**Reply by xFile3160** [reply]
FWIW: https://github.com/xFile3160/bazel-central-registry

added:
halide
mnn
opencl-headers
opencl-hpp
node-addon-api
node-api-headers
flatbuffers (but it got added recently as well last time I checked).

**Comment by limdor** [comment]
[glew](https://glew.sourceforge.net/)

**Comment by limdor** [comment]
[assimp](https://github.com/assimp/assimp)

**Comment by photex** [comment]
Would be great to have an [ispc](https://ispc.github.io/) toolchain.

**Reply by Vertexwahn** [reply]
https://github.com/Vertexwahn/rules_ispc (does not support bzlmod yet)

**Comment by photex** [comment]
Oh! Skia would be lovely. And SDL2.

**Comment by muhammadfarooq1415** [comment]
[ortools](https://github.com/google/or-tools)

**Reply by muhammadfarooq1415** [reply]
[issue](https://github.com/bazelbuild/bazel-central-registry/issues/1404)

**Comment by AustinSchuhBRT** [comment]
gRPC doesn't really appear supported.  There are patches to do it, but they didn't work for me.

**Reply by meteorcloudy** [reply]
Can you please comment on https://github.com/grpc/grpc/issues/35271 to explain what issues you encountered?

**Comment by alexeagle** [comment]
@meteorcloudy I think it may be time to archive this discussion?

**Reply by meteorcloudy** [reply]
SG, most modules requested here are in BCR, if there are more please file issues at https://github.com/bazelbuild/bazel-central-registry/issues! Closing this one now.

**Comment by sluongng** [comment]
cc: @alexeagle since this issue is closed, might need to update the URL in https://opencollective.com/bazel-rules-authors-sig/projects/bazel-central-registry

**Reply by alexeagle** [reply]
fixed, thx

**Comment by mapleqin** [comment]
ffmpeg
skia
